runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -653,19 +650,34 @@ private void processElement(Map<W, W> windowToMergeResult, WindowedValue<InputT>
     // end-of-window time to be a signal to garbage collect.
     public final boolean isGarbageCollection;
 
-    EnrichedTimerData(
-        TimerData timer,
+    WindowActivation(
         ReduceFn<K, InputT, OutputT, W>.Context directContext,
         ReduceFn<K, InputT, OutputT, W>.Context renamedContext) {
-      this.timestamp = timer.getTimestamp();
       this.directContext = directContext;
       this.renamedContext = renamedContext;
       W window = directContext.window();
-      this.isEndOfWindow = TimeDomain.EVENT_TIME == timer.getDomain()
-          && timer.getTimestamp().equals(window.maxTimestamp());
-      Instant cleanupTime = LateDataUtils.garbageCollectionTime(window, windowingStrategy);
+
+      // The output watermark is before the end of the window if it is either unknown
+      // or it is known to be before it. If it is unknown, that means that there hasn't been
+      // enough data to advance it.
+      boolean outputWatermarkBeforeEOW =
+              timerInternals.currentOutputWatermarkTime() == null
+          || !timerInternals.currentOutputWatermarkTime().isAfter(window.maxTimestamp());
+
+      // The "end of the window" is reached when the local input watermark (for this key) surpasses
+      // it but the local output watermark (also for this key) has not. After data is emitted and
+      // the output watermark hold is released, the output watermark on this key will immediately
+      // exceed the end of the window (otherwise we could see multiple ON_TIME outputs)
+      this.isEndOfWindow =
+          timerInternals.currentInputWatermarkTime().isAfter(window.maxTimestamp())
```

Comment:
It is a good point - the oft-state logic is that "the ON_TIME output is the unique output where the (local) input WM > EOW and the (local) output WM <= EOW". But that only applies when the local output WM is tracked explicitly. I think the watermark hold here is very nearly this, but we don't use it quite that way.

I think this still works because:

 - We simply statefully track whether we have emitted a non-`EARLY` pane, so we'll never create more than one `ON_TIME` pane.
 - If allowed lateness is zero, then we drop data based on the input watermark, so there will never be more than out output with the default trigger.
 - If allowed lateness is nonzero, then the default trigger allows late outputs and the state we keep ensures these are right. For other triggers, such as a straight EOW trigger, the fact that the trigger closes ensures that it does not produce more output.

The logic would be more straightforward if we explicitly tracked the watermark for just this key and let it zoom ahead of the output watermark for the entire output collection.